core/src/main/scala/io/projectglow/vcf/BigVCFDatasource.scala

```diff
@@ -44,11 +45,20 @@ object BigVCFDatasource extends HlsUsageLogging {
 
     val dSchema = data.schema
     val rdd = data.queryExecution.toRdd
+
     val nParts = rdd.getNumPartitions
-    val serializableConf = new SerializableConfiguration(
-      VCFFileFormat.hadoopConfWithBGZ(data.sparkSession.sparkContext.hadoopConfiguration)
-    )
-    val writerFactory = new VCFOutputWriterFactory(options)
+    val conf = VCFFileFormat.hadoopConfWithBGZ(data.sparkSession.sparkContext.hadoopConfiguration)
+    val serializableConf = new SerializableConfiguration(conf)
+    val firstNonemptyPartition =
+      rdd.mapPartitions(iter => Iterator(iter.nonEmpty)).collect.indexOf(true)
+
+    if (firstNonemptyPartition == -1 && (!options.contains(VCFFileWriter.VCF_HEADER_KEY) ||
+      options
+        .get(VCFFileWriter.VCF_HEADER_KEY)
+        .contains(VCFFileWriter.INFER_HEADER))) {
+      rdd.unpersist()
```

Comment:
You don't need this if you handle errors properly in `BigFileDatasource`

Reply:
Whoops, I don't think the RDD was cached to begin with.